File: Java/src/com/eigenvektor/tree/Tree.java
/*
 *  Interface for a tree.
 *  Copyright (C) 2013 Michael Thorsley
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package com.eigenvektor.tree;

import java.util.Collection;
import java.util.Iterator;

/**
 * Specification for a tree.  The number of children of each node is unspecified.  The edges of the tree
 * are implicitly weighted.
 *
 * @param <T> The node type of the tree.  This in general should have equals() and hashCode() implemented
 * correctly.
 */
public interface Tree<T> {

	/**
	 * Gets the root of the tree.
	 * 
	 * @return The root node of the tree.
	 */
	T getRoot();
	
	/**
	 * Tells if an object is a node of this tree.
	 * 
	 * @param x The object to consider.
	 * @return true iff <i>x</i> is in the tree.
	 */
	public boolean isNode(final T x);
	
	/**
	 * Gets the weight of a node.
	 * 
	 * @param x The node to consider.
	 * @return The weight of <i>x</i>.
	 */
	public double getWeight(final T x);
	
	/**
	 * Gets the parent of a node.
	 * 
	 * @param x The node to consider.
	 * @return The parent of <i>x</i>, or <i>null</i> if <i>x</i> is the root.
	 */
	public T getParent(final T x);
	
	/**
	 * Gets the children of a node.
	 * 
	 * @param x
	 * @return The children of <i>x</i>.  If the children's order is important to the 
	 * implementation, returns an ordered collection.  Return value is not necessarily
	 * mutable.
	 */
	Collection<T> getChildren(final T x);

	
	/**
	 * Gets an iterator that traverses the tree depth first.  If the children of nodes are ordered, their
	 * subtrees are traversed in order.
	 * 
	 * @return An iterator that does that iteration in depth first order.
	 */
	Iterator<T> getDepthFirstIterator();
	
	/**
	 * Gets an iterator that traverses the tree breadth first.  If the children of nodes are ordered, they
	 * are traversed in order.
	 * 
	 * @return An iterator that does that iteration in breadth first order.
	 */
	Iterator<T> getBreadthFirstIterator();
}

File: Scala/src/com/eigenvektor/graph/Tree.scala
/*
 *  Trait for a tree.
 *  Copyright (C) 2013 Michael Thorsley
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package com.eigenvektor.graph

/**
 * A trait for all trees.
 */
trait Tree[E] extends Flow[E] {
  
  /** The root of the tree */
  def root:E

  /** The roots of the tree.  There is only one. */
  def roots = Set(root)
  
  /** The children of the tree. */
  def children:List[Tree[E]]
  
  /** Tells if this is a leaf */
  def isLeaf:Boolean = children.isEmpty
  
  /** Applies a function to each node in preorder
   *  
   *  @param f the function to apply.
   */
  def preorderForeach(f: E => Unit):Unit = {
    f(root)
    children.foreach(_.preorderForeach(f))
  }
  
  /** Applies a function to each node in preorder
   *  
   *  @param f the function to apply.
   */
  def postorderForeach(f: E => Unit):Unit = {
    children.foreach(_.postorderForeach(f))
    f(root)
  }
  
  /** Adds another tree to this as the first subtree */
  def +(t:Tree[E]):Tree[E]
}

object Tree {
  
  /** Creates a tree with kids */
  def apply[E](root:E, kids:Tree[E]*) = new DefaultTree(root, kids.toList)
  
  /** Unapply for patter matching */
  def unapply[E](t:Tree[E]):Option[(E, List[Tree[E]])] = Some(t.root, t.children)
}